src/NzbDrone.Core/DecisionEngine/Specifications/UpgradableSpecification.cs

```diff
@@ -57,9 +90,48 @@ private bool IsPreferredWordUpgradable(int currentScore, int newScore)
             return newScore > currentScore;
         }
 
+        public bool IsUpgradable(QualityProfile qualityProfile, LanguageProfile languageProfile, List<QualityModel> currentQualities, List<Language> currentLanguages, int currentScore, QualityModel newQuality, Language newLanguage, int newScore)
+        {
+            if (IsQualityUpgradable(qualityProfile, currentQualities, newQuality))
+            {
+                return true;
+            }
+
+            foreach (var quality in currentQualities)
+            {
+                if (new QualityModelComparer(qualityProfile).Compare(newQuality, quality) < 0)
+                {
+                    _logger.Debug("Existing item has better quality, skipping");
+                    return false;
+                }
+            }
+
+            if (IsLanguageUpgradable(languageProfile, currentLanguages, newLanguage))
+            {
+                return true;
+            }
+
+            foreach (var language in currentLanguages)
+            {
+                if (new LanguageComparer(languageProfile).Compare(newLanguage, language) < 0)
+                {
+                    _logger.Debug("Existing item has better language, skipping");
+                    return false;
+                }
+            }
+
+            if (!IsPreferredWordUpgradable(currentScore, newScore))
+            {
+                _logger.Debug("Existing item has a better preferred word score, skipping");
+                return false;
+            }
+
+            return true;
+        }
+
         public bool IsUpgradable(QualityProfile qualityProfile, LanguageProfile languageProfile, QualityModel currentQuality, Language currentLanguage, int currentScore, QualityModel newQuality, Language newLanguage, int newScore)
         {
```

Comment:
Rather than duplicating the logic here can we just construct singleton lists for `currentQuality` and `currentLanguage` and use the function above

Reply:
Yeah that was my first take at it, got lazy on changing the instances. I agree it will be cleaner.